lib/rules/html-indent.js

```diff
@@ -0,0 +1,1228 @@
+}
+
+/**
+ * Get the last element.
+ * @param {Array} xs The array to get the last element.
+ * @returns {any|undefined} The last element or undefined.
+ */
+function last (xs) {
+  return xs.length === 0 ? undefined : xs[xs.length - 1]
+}
+
+/**
+ * Check whether the node is at the beginning of line.
+ * @param {Node} node The node to check.
+ * @param {number} index The index of the node in the nodes.
+ * @param {Node[]} nodes The array of nodes.
+ * @returns {boolean} `true` if the node is at the beginning of line.
+ */
+function isBeginningOfLine (node, index, nodes) {
+  if (node != null) {
+    for (let i = index - 1; i >= 0; --i) {
+      const prevNode = nodes[i]
+      if (prevNode == null) {
+        continue
+      }
+
+      return node.loc.start.line !== prevNode.loc.end.line
+    }
+  }
+  return false
+}
+
+/**
+ * Creates AST event handlers for html-indent.
+ *
+ * @param {RuleContext} context - The rule context.
+ * @returns {object} AST event handlers.
+ */
+function create (context) {
+  const options = parseOptions(context.options[0], context.options[1] || {})
+  const sourceCode = context.getSourceCode()
+  const template = context.parserServices.getTemplateBodyTokenStore && context.parserServices.getTemplateBodyTokenStore()
+  const offsets = new Map()
+
+  /**
+   * Set offset to the given tokens.
+   * @param {Token|Token[]} token The token to set.
+   * @param {number} offset The offset of the tokens.
+   * @param {Token} baseToken The token of the base offset.
+   * @returns {void}
+   */
+  function setOffset (token, offset, baseToken) {
+    assert(baseToken != null, "'baseToken' should not be null or undefined.")
+
+    if (Array.isArray(token)) {
+      for (const t of token) {
+        offsets.set(t, {
+          baseToken,
+          offset,
+          baseline: false,
+          expectedIndent: undefined
+        })
+      }
+    } else {
+      offsets.set(token, {
+        baseToken,
+        offset,
+        baseline: false,
+        expectedIndent: undefined
+      })
+    }
+  }
+
+  /**
+   * Set baseline flag to the given token.
+   * @param {Token} token The token to set.
+   * @returns {void}
+   */
+  function setBaseline (token, hardTabAdditional) {
+    const offsetInfo = offsets.get(token)
+    if (offsetInfo != null) {
+      offsetInfo.baseline = true
+    }
+  }
+
+  /**
+   * Get the first and last tokens of the given node.
+   * If the node is parenthesized, this gets the outermost parentheses.
+   * @param {Node} node The node to get.
+   * @param {number} [borderOffset] The least offset of the first token. Defailt is 0. This value is used to prevent false positive in the following case: `(a) => {}` The parentheses are enclosing the whole parameter part rather than the first parameter, but this offset parameter is needed to distinguish.
+   * @returns {{firstToken:Token,lastToken:Token}} The gotten tokens.
+   */
+  function getFirstAndLastTokens (node, borderOffset) {
+    borderOffset |= 0
```

Comment:
could you explain a little? :)

Reply:
its js `magic` assign 0 if value is non numeric

Reply:
hah.. got it! amazing! lol~

Reply:
Yes, it is "casting to signed integer".

Bitwise operators (`&`, `|`, "^", `~`, `<<`, `>>`, and `>>>`) casts both operands to 32bits signed integer, then the logical OR with `0` does nothing. As the result, the `x | 0` expression casts `x` to 32bits signed integer.

Reply:
@mysticatea like i said `magic`